### ISSUE TITLE
Reproducing a stack overflow example

### ISSUE THREAD

**Issue by tboerstad** [OverLordGoldDragon/see-rnn#10]
Thank you very much for creating see-rnn! 
I've been working with RNNs for years, 
yet this kind of visualization was eye opening for me. 

In the bottom of your Stack Overflow post, you are visualizing gradients which have a near constant magnitude for each timestep. How can I reproduce something similar in the provided `rnn_sandbox`?

My gradients always vanish towards zero for the earlier timesteps. 

Link to stack overflow:
https://stackoverflow.com/questions/48714407/rnn-regularization-which-component-to-regularize

**Comment by OverLordGoldDragon** [comment]
Glad you found it useful.

Fine question; I'll possibly dedicate a Q&A to it on SO. But briefly: I'm fairly sure careful _weight decay_ is responsible. I haven't found a concrete guideline anywhere as to how to tune them besides "see what works", but the pitfall of that is lack of _orthogonality_; `1e-4` may work best in network X, but not Y. A _subnetwork-oriented_ approach is required - so I made my own: _set a target l2-norm, channel-wise_. A usable script, along further explanation, below.

Note, though, the rest of that SO answer is relevant; it's important to not _wrongly_ regularize, else it can bottleneck the entire layer (e.g. large `dropout`). Also, I'd be careful in trusting the default optimizers with the subnetwork principle, as l2-penalty gradients are tied to second moments, which are [unreliable](https://github.com/OverLordGoldDragon/keras-adamw/issues/18#issuecomment-546736493). Lastly, may find interest in [this discussion](https://github.com/tensorflow/tensorflow/issues/35724#issuecomment-577251311) on `recurrent_dropout`.

<hr>

 **Improved gradient persistence along time**; with a bounded activation like `sigmoid` or `tanh` in the hidden-to-hidden transformation, large weights are likelier to yield larger pre-activations, which drives _activations_ closer to saturation, diminishing the gradient. The effect is amplified upon recurrence - especially for long sequences, _inhibiting learning long-term dependencies_.

<hr>

```python
def inspect_weights_l2(model, names=['dense', 'lstm'], axis=-1):
    def _get_l2(w, axis=-1):
        axis = axis if axis != -1 else len(w.shape) - 1
        reduction_axes = tuple([ax for ax in range(len(w.shape)) if ax != axis])
        return np.sqrt(np.sum(np.square(w), axis=reduction_axes))
    
    def _print_layer_l2(layer, idx, axis=-1):
        W = layer.get_weights()
        l2_all = []
        txt = "{} "
        
        for w in W:
            txt += "{:.4f}, {:.4f} -- "
            l2 = _get_l2(w, axis)
            l2_all.extend([l2.max(), l2.mean()])
        txt = txt.rstrip(" -- ")
        
        print(txt.format(idx, *l2_all))

    names = [names] if isinstance(names, str) else names

    for idx, layer in enumerate(model.layers):
        if any([name in layer.name.lower() for name in names]):
            _print_layer_l2(layer, idx, axis=axis)
```

**Comment by tboerstad** [comment]
Thank you for your answer. 

I used your AdamW optimizer and modified the weight decays, now my gradients no longer vanish 👍  
It's not picture perfect like your SO answer, but it works
<img width="1091" alt="gradients" src="https://user-images.githubusercontent.com/4872288/75267794-fad42580-57f6-11ea-8bd6-80c1b47ee10f.png">

**Comment by tboerstad** [comment]
Thank you for your answer. 

I used your AdamW optimizer, and now my gradients no longer vanish. 
It's not picture perfect like your SO answer, but it's a lot better than what I had
<img width="1091" alt="gradients" src="https://user-images.githubusercontent.com/4872288/75267794-fad42580-57f6-11ea-8bd6-80c1b47ee10f.png">

**Comment by OverLordGoldDragon** [comment]
@tboerstad Nicely done; yeah, my case is a rare one, as noted in the answer, every relevant hyperparameter was carefully tuned.

**Comment by OverLordGoldDragon** [comment]
@tboerstad Full [SO example](https://stackoverflow.com/questions/61481921/how-to-set-and-track-weight-decays) done; good inspiration

**Comment by tboerstad** [comment]
Thank you for providing your time and value to the community. 
I have not found this kind of content anywhere until now.

**Comment by OverLordGoldDragon** [comment]
@tboerstad Thanks, glad it's of value
